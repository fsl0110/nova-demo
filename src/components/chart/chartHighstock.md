#### Chart without data

```jsx
import moment from "moment";

const testProps = {
  location: "reports",
  fetchHandler: () => null,
  chartOptions: {
    chart: {
      height: 350,
      style: {
        fontFamily:
          "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
      }
    },
    credits: {
      enabled: false
    },
    loading: {
      hideDuration: 500,
      showDuration: 300
    },
    title: {
      text: ""
    },
    yAxis: {
      title: {
        text: null
      }
    },
    xAxis: {
      type: "datetime",
      labels: {
        formatter: function() {
          return moment(this.value).format("MMM YYYY");
        }
      },
      minTickInterval: moment.duration(1, "month").asMilliseconds()
    },
    series: [{ type: "line", data: [null] }]
  }
};

<ChartHighstock {...testProps} />;
```

#### Chart with error

```jsx
const testProps = {
  location: "initiators",
  hasError: true,
  fetchHandler: () => null,
  chartOptions: {
    series: [{ type: "line", data: [0, 0] }]
  }
};

<ChartHighstock {...testProps} />;
```

#### Chart while loading

```jsx
const testProps = {
  location: "initiators",
  isLoading: true,
  fetchHandler: () => null,
  chartOptions: {}
};

<ChartHighstock {...testProps} />;
```
