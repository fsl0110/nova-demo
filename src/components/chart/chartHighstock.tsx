import React, { HTMLAttributes, PureComponent } from "react";
import Highcharts from "highcharts/highstock";
import HighchartsReact from "highcharts-react-official";
import classNames from "classnames";
import { Error } from "..";
require("highcharts-no-data-to-display")(Highcharts);

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** Is loading data. */
  isLoading?: boolean;
  /** Has an error. */
  hasError?: boolean;
  /** Options and Data of the chart */
  chartOptions: Highcharts.Options;
  /** Action for fetching data. */
  fetchHandler: (term?: string) => void;
}

export class ChartHighstock extends PureComponent<Props> {
  private chartInstance: any;

  constructor(props: Props) {
    super(props);

    this.afterChartCreated = this.afterChartCreated.bind(this);
  }

  afterChartCreated(chart: any) {
    this.chartInstance = chart;
  }

  componentDidMount() {
    this.props.fetchHandler();
  }

  render() {
    const {
      isLoading,
      hasError,
      chartOptions,
      fetchHandler,
      className
    } = this.props;

    if (!hasError && this.chartInstance && isLoading) {
      this.chartInstance.showLoading();
    }
    if (!hasError && this.chartInstance && !isLoading) {
      this.chartInstance.hideLoading();
    }

    return (
      <div className={classNames(className)} data-test="ChartHighstock">
        {!hasError ? (
          <HighchartsReact
            highcharts={Highcharts}
            options={chartOptions}
            constructorType={"stockChart"}
            callback={this.afterChartCreated}
          />
        ) : (
          <Error fetchData={fetchHandler} isLoading={isLoading} />
        )}
      </div>
    );
  }
}

export default ChartHighstock;
