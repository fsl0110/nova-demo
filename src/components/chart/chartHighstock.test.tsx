import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { ChartHighstock, Props } from "./chartHighstock";
import moment from "moment";

const testProps: Props = {
  location: "reports",
  fetchHandler: jest.fn(),
  chartOptions: {
    chart: {
      height: 350,
      style: {
        fontFamily:
          "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
      }
    },
    credits: {
      enabled: false
    },
    loading: {
      hideDuration: 500,
      showDuration: 300
    },
    title: {
      text: ""
    },
    yAxis: {
      title: {
        text: null
      }
    },
    xAxis: {
      type: "datetime",
      labels: {
        formatter: function() {
          return moment(this.value).format("MMM YYYY");
        }
      },
      minTickInterval: moment.duration(1, "month").asMilliseconds()
    },
    series: [{ type: "line", data: [null] }]
  }
};

describe("<ChartHighstock>", () => {
  it("renders without crashing", () => {
    shallow(<ChartHighstock {...testProps} />);
  });

  it("renders with error", () => {
    const newTestProps = { ...testProps, hasError: true };
    const mock = shallow(<ChartHighstock {...newTestProps} />);
    expect(toJson(mock)).toMatchSnapshot();
  });
});
