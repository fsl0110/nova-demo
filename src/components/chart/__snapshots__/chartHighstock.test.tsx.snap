// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartHighstock> renders with error 1`] = `
<div
  className=""
  data-test="ChartHighstock"
>
  <Error
    fetchData={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  />
</div>
`;
